keybinding concept (#994)
ElektraInitiative/PermaplanT: doc/guidelines/frontend-keyhandling.md

# Frontend Keybindings

This document gives guidelines on how to implement keybindings in the frontend.
Our key handling solution is designed to address the following key constraints:

- **Developer-friendly**: Developers should get an overview where which key binding is assigned.
- **UX**: Collisions between shortcuts which trigger multiple actions at once should be avoided.

> ### Important notes for Konva Key handling

Reviewer: There are several change requests still open and also rebase is required.

> ### Important notes for Konva Key handling

Author: oh sorry for some reason i didn't notice that there a further change requests. i will take care of them now

> - since keys cannot be bound directly on konva elements, keyhandlers have to check if the corresponding layer is active to avoid collisions

Author: i added a checkbox/todo here: https://github.com/ElektraInitiative/PermaplanT/issues/1031.
For implementing this feature I need to check if a layer is active. After implementing this, I will provide specific guidelines on how to do it

> - since keys cannot be bound directly on konva elements, keyhandlers have to check if the corresponding layer is active to avoid collisions

Reviewer: Ok, the macOS thing, if it is not trivial/obvious, can also be done with #1031.

- if keys should only be active if map is focused, handlers have to be bound to canvas section of the map component

### Keybinding Configuration

The file 'keybindings.json' contains all keybindings that are used in the application and serves the following characteristics:

- keybindings can easily be changed by editing the file
- developers have an overview of all keybindings.
- there are different sections for keybindings:
  - global: keybindings that are globally active
  - scopes for specific layers: keybindings that are only active if the corresponding layer is active. this scopes should be named according to the layer name

>   - _just_for_documentation_: keybindings that are only used for documentation purposes and are not meant to get configured

Author: i added an example to make clear what i meant

- modifier keys are supported. (e.g. Ctrl+k)
- multiple keybindings can be assigned to one action

Structure:

```
{
  "global": {

>     "<<action_name>>": ["<<Key-Binding>>, ...]"

Reviewer 2: I'd suggest swapping the key and the value here if it is not too much effort:
"<<key_binding>>": "<<action_name>>"
The reason is that I think it is easier to look up the corresponding action within the event handler: e.g. config.global['Escape'] to get the corresponding action.
It is also easier to check if there are keybindings assigned multiple times as the json is not valid anymore.

Reviewer: I also suggested that already but everyone else found it more clear as it is proposed now. A check if key-bindings are used several times should be quite simple in both cases.

> config.global['Escape']

This looks to me to a reason to keep it as proposed. The code shouldn't bother about which keybinding is used, otherwise you couldn't simply make changes in the JSON. Or did I misunderstand?

Reviewer 2: Fair enough. I don't have a strong opinion about it, it is just more intuitive for me as in my experience keybindings are often defined with binding -> action in my experience.
For example vim config, sxhkd config but that doesn't necessarily mean that it is better for us :)
It is not hard to check for duplicates in the config file but my point was that you have to do it.
When the keybinding is the key we have an invalid json as soon as the keybinding is used twice.
On the other hand it might be quicker to look up the bindings from the config with the action as key so we can leave it like that :)

  }
  "plants_layer": {
    "exitPlantingMode": ["Escape"]
  }
  ...
  "_just_for_documentation_": {
    "clearSearch": ["Escape"}
  }
}
```

### Keyhandling on Focused Elements

- for components that provide keylisteners and are usually focused on use (e.g. input fields)
- method: add scope to JSON, bind a keyhandler and execute action according to configuration
- if actions of parent components shouldn't be triggered, event propagation must be stopped

Example

```
    const handleKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
      const action = getActionNameFromKeyEvent(<scope>, event)
      if(action === "doSomething"){
        doSomething();
        event.stopPropagation();
      }
    };
```

### Keyhandling independent of focus

- used if keylistener should be triggered independent of focused HTML element
- method: use custom keybinding hook and pass handlers according to config

Example:

```
const keyHandlerActions: Record<string, () => void> = {
  exitPlantingMode: () => {
    exitPlantingMode();
  },
};

> useKeyHandlers(createKeyHandlersFromConfig("planting_layer", keyHandlerActions));

Reviewer: `useKeyHandlers` is a function from you? Please introduce it here shortly or at least add comments in the example.

In #1016 only createKeyBindingsAccordingToConfig exists.

Please also make sure the snippets are formatted correctly.

Author: useKeyHandlers is the custom hook we introduced in a earlier PR that is already merged. i will add a comment.

# Frontend Keybindings

This document gives guidelines on how to implement keybindings in the frontend.

### Keybinding Configuration

> The file 'keybindings.json' contains all keybindings that are used in the application and serves the following characteristics:

Reviewer: ```suggestion
The file `config/keybindings/keybindings.json` contains all keybindings that are used in the application and serves the following characteristics:
```

Please focus a bit more

> 

Reviewer: ```suggestion

- for all features add commonly-used and vim-like bindings
- we generally try to provide several keybindings for the same functionality, to make the functionality easier accessible
```

> - keybindings can easily be changed by editing the file
> - developers have an overview of all keybindings

Reviewer: These are not really guidelines?

```suggestion
```

- there are different sections for keybindings:

>   - global: keybindings that are globally active

Reviewer: (Also missing at other places)

```suggestion
  - `global`: keybindings that are globally active
```

  - Scopes for specific layers:
    Keybindings that are only active if the corresponding layer is active.
    These scopes should be named according to the layer name, postfixed with `_layer`.
  - _just_for_documentation_:
    Keybindings that are only used for documentation purposes since the keybinding is hardcoded in the component.
    For example, the search field can be cleared by pressing the Escape key, which is typically a native browser action and should not be configurable in the keybindings file.
- modifier keys are supported (Shift, Alt, Ctrl, Meta)

>   - they can be used in combination with other keys by using the '+' sign (e.g. 'Alt+P', 'Shift+A')

Reviewer: How do these map on Mac OS?
Can you maybe introduce a wildcard modifier key so that it automatically works on Mac OS, too?

Author: the react key event supports the four mentioned modifier keys (Shift, Alt, Ctrl, Meta). 
The keys should be mapped like this:

- event.ctrlKey: Represents whether the Control key (Ctrl) is pressed.
- event.shiftKey: Represents whether the Shift key is pressed.
- event.altKey: Represents whether the Alt key is pressed (Option key on macOS).
- event.metaKey: Represents whether the Meta key (Command key on macOS) is pressed.

Unfortunately, i cannot test if it is really working like this. I created a left over issue: https://github.com/ElektraInitiative/PermaplanT/issues/1057

- multiple keybindings can be assigned to one action

Structure:

```json
{
  "global": {
    "<<action_name>>": ["<<Key-Binding>>", "..."]
  },
  "base_layer": {
    "<<action_name>>": ["<<Key-Binding>>", "..."]
  },
  "plants_layer": {
    "exitPlantingMode": ["Escape"]
  },
  "_just_for_documentation_": {
    "clearSearch": ["Escape"]
  }
}
```

> ### Keyhandling on Focused Elements

Reviewer: First describe what should be done in any case. Ideally add a step-by-step description what someone should do to properly add a keybinding. Then add info about special cases.

Author: in my opinion the subsections are not describing special cases. there are just two different ways we have to handle keyhandling for either focused elements or keybindings that should be available independent of focus. 
i created a more detailed step-by-step description for both variants.
